File: clo2plt/document.py
"""Load a CLO 3D marker PDF into classified, plotter-ready geometry."""

import re
from dataclasses import dataclass, field

from . import classify
from . import geometry as g
from .content import Interpreter, parse_to_unicode
from .pdfread import PdfError, load

MM_PER_POINT = 25.4 / 72.0


@dataclass
class Piece:
    name: str
    strokes: list = field(default_factory=list)
    labels: list = field(default_factory=list)


@dataclass
class Marker:
    path: str
    page_w: float          # mm
    page_h: float          # mm
    pieces: list
    strokes: list
    labels: list
    unknown_colors: dict
    fills_dropped: int
    curves: int
    lines: int
    producer: str = ""

    @property
    def bbox(self):
        boxes = [s.points for s in self.strokes]
        return g.bbox(boxes)

    def counts(self):
        tally = {k: 0 for k in classify.ORDER}
        for s in self.strokes:
            tally[s.kind] += 1
        tally[classify.LABEL] = len(self.labels)
        return tally


def read(path, tolerance=0.05, pieces=None):
    """Parse `path` into a Marker. `pieces` optionally filters by piece name."""
    pdf = load(path)
    page = _first_page(pdf)
    body = pdf.raw(page)

    media = pdf.numbers(body, "MediaBox")
    if not media or len(media) != 4:
        raise PdfError("page has no usable /MediaBox")

    # Land everything in millimetres: the page `cm` maps content units to
    # points, and this converts points to mm. Tolerances and output are mm
    # throughout from here on.
    unit = (MM_PER_POINT, 0.0, 0.0, MM_PER_POINT,
            -media[0] * MM_PER_POINT, -media[1] * MM_PER_POINT)
    interp = Interpreter(pdf, tolerance=tolerance, to_unicode=_to_unicode(pdf))
    result = interp.run_page(page, unit)

    # Page content is millimetres; the page-level `cm` supplies the scale that
    # maps them onto the MediaBox's points. Deriving it (rather than assuming)
    # means a differently-scaled export is caught below instead of mis-sized.
    page_w = (media[2] - media[0]) * MM_PER_POINT
    page_h = (media[3] - media[1]) * MM_PER_POINT

    unknown = {}
    for s in result.strokes:
        s.kind = classify.line_type(s.rgb)
        if s.kind == classify.UNKNOWN:
            key = tuple(round(v, 6) for v in s.rgb)
            unknown[key] = unknown.get(key, 0) + 1

    wanted = None
    if pieces:
        wanted = {p.strip().casefold() for p in pieces if p.strip()}
        result.strokes = [s for s in result.strokes if s.piece.casefold() in wanted]
        result.labels = [l for l in result.labels if l.piece.casefold() in wanted]
        found = {s.piece.casefold() for s in result.strokes}
        missing = wanted - found
        if missing:
            raise PdfError(
                "no such piece(s): " + ", ".join(sorted(missing)) + ". Use --list."
            )

    by_name = {}
    for s in result.strokes:
        by_name.setdefault(s.piece, Piece(s.piece)).strokes.append(s)
    for l in result.labels:
        by_name.setdefault(l.piece, Piece(l.piece)).labels.append(l)

    producer = ""
    m = re.search(rb"/Producer\s*\(([^)]*)\)", pdf.data)
    if m:
        producer = m.group(1).decode("latin-1")

    return Marker(
        path=path,
        page_w=page_w,
        page_h=page_h,
        pieces=list(by_name.values()),
        strokes=result.strokes,
        labels=result.labels,
        unknown_colors=unknown,
        fills_dropped=result.fills_dropped,
        curves=result.curves,
        lines=result.lines,
        producer=producer,
    )


def _first_page(pdf):
    for m in re.finditer(rb"(\d+)\s+0\s+obj", pdf.data):
        num = int(m.group(1))
        body = pdf.raw(num)
        if re.search(rb"/Type\s*/Page[^s]", body):
            return num
    raise PdfError("no page object found")


def _to_unicode(pdf):
    """Merge every ToUnicode CMap in the file; CLO emits a single shared font."""
    mapping = {}
    for m in re.finditer(rb"/ToUnicode\s+(\d+)\s+\d+\s+R", pdf.data):
        try:
            mapping.update(parse_to_unicode(pdf.stream(int(m.group(1)))))
        except PdfError:
            continue
    return mapping

File: site/js/document.js
// Load a CLO 3D marker PDF into classified geometry. Mirrors clo2plt/document.py.

import * as classify from "./classify.js";
import * as g from "./geometry.js";
import { Interpreter, parseToUnicode } from "./content.js";
import { PdfError, load } from "./pdfread.js";

export const MM_PER_POINT = 25.4 / 72;

export async function read(bytes, { tolerance = 0.05, pieces = null } = {}) {
  const pdf = await load(bytes);
  const page = firstPage(pdf);
  const body = pdf.raw(page);

  const media = pdf.numbers(body, "MediaBox");
  if (!media || media.length !== 4) {
    throw new PdfError("This page has no usable /MediaBox.");
  }

  // Land everything in millimetres: the page `cm` maps content units to
  // points, and this converts points to mm.
  const unit = [
    MM_PER_POINT, 0, 0, MM_PER_POINT,
    -media[0] * MM_PER_POINT, -media[1] * MM_PER_POINT,
  ];
  const interp = new Interpreter(pdf, tolerance, toUnicode(pdf));
  interp.runPage(page, unit);

  const pageW = (media[2] - media[0]) * MM_PER_POINT;
  const pageH = (media[3] - media[1]) * MM_PER_POINT;

  const unknown = new Map();
  for (const s of interp.strokes) {
    s.kind = classify.lineType(s.rgb);
    if (s.kind === classify.UNKNOWN) {
      const key = s.rgb.map((v) => v.toFixed(6)).join(" ");
      unknown.set(key, (unknown.get(key) || 0) + 1);
    }
  }

  let strokes = interp.strokes;
  let labels = interp.labels;
  if (pieces && pieces.length) {
    const wanted = new Set(pieces.map((p) => p.trim().toLowerCase()).filter(Boolean));
    strokes = strokes.filter((s) => wanted.has(s.piece.toLowerCase()));
    labels = labels.filter((l) => wanted.has(l.piece.toLowerCase()));
    const found = new Set(strokes.map((s) => s.piece.toLowerCase()));
    const missing = [...wanted].filter((w) => !found.has(w));
    if (missing.length) throw new PdfError(`No such piece: ${missing.join(", ")}`);
  }

  const byName = new Map();
  for (const s of strokes) {
    if (!byName.has(s.piece)) byName.set(s.piece, { name: s.piece, strokes: [], labels: [] });
    byName.get(s.piece).strokes.push(s);
  }
  for (const l of labels) {
    if (!byName.has(l.piece)) byName.set(l.piece, { name: l.piece, strokes: [], labels: [] });
    byName.get(l.piece).labels.push(l);
  }

  const prod = /\/Producer\s*\(([^)]*)\)/.exec(pdf.text);

  const marker = {
    pageW, pageH,
    pieces: [...byName.values()],
    strokes, labels,
    unknownColors: unknown,
    fillsDropped: interp.fillsDropped,
    curves: interp.curves,
    lines: interp.lines,
    producer: prod ? prod[1] : "",
  };
  marker.bbox = g.bbox(strokes.map((s) => s.points));
  marker.counts = counts(marker);
  return marker;
}

export function counts(marker) {
  const tally = {};
  for (const k of classify.ORDER) tally[k] = 0;
  for (const s of marker.strokes) tally[s.kind] += 1;
  tally[classify.LABEL] = marker.labels.length;
  return tally;
}

function firstPage(pdf) {
  for (const num of pdf.offsets.keys()) {
    if (/\/Type\s*\/Page[^s]/.test(pdf.raw(num))) return num;
  }
  throw new PdfError("No page object found in this PDF.");
}

function toUnicode(pdf) {
  const mapping = new Map();
  for (const m of pdf.text.matchAll(/\/ToUnicode\s+(\d+)\s+\d+\s+R/g)) {
    const num = Number(m[1]);
    if (!pdf.hasStream(num)) continue;
    for (const [k, v] of parseToUnicode(pdf.stream(num))) mapping.set(k, v);
  }
  return mapping;
}
